Tidy up the Graph component (part #2)

kinow opened the pull request:
These changes partially address #74 and #381, see also #302 

Initial view of the Graph component when using this branch:

![image](https://user-images.githubusercontent.com/304786/72981116-e93adf00-3e40-11ea-8592-e27a6e02222d.png)

After dismissing the warning, and expanding the layout options:

![image](https://user-images.githubusercontent.com/304786/72981161-fce64580-3e40-11ea-831c-5fe3ed58f0ce.png)

Now left with a frozen graph, and dagre engine. While right has dismissed the warning, hierarchical layout, layout options hidden, and it is showing live data.

![image](https://user-images.githubusercontent.com/304786/72981242-361eb580-3e41-11ea-8520-c2baf74ffec7.png)

Tests above were done with Firefox 72.0.1 (64 bit) on Ubuntu LTS, using my home laptop.

@hjoliver it kinda works with 60.8. Will take another look tomorrow, but I suspect it's the Lumino widget and how it is handling updates and resizing.

For me, the `/#/graph/families2` view worked fine.

![image](https://user-images.githubusercontent.com/304786/72981943-d5907800-3e42-11ea-87e4-cd356d2f313c.png)

But got the same issue reported by @hjoliver when using Lumino.

![image](https://user-images.githubusercontent.com/304786/72981960-e214d080-3e42-11ea-9cbe-998f9e9d6059.png)

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Does not need tests (why? it's simply deleting or moving code).
- [x] No change log entry required (why? e.g. invisible to users).
- [x] No documentation update required.

kinow commented:
Notes to self:

- I think multiple Graph components with Lumino is broken due to the ID of the DIV container being used... maybe refs will fix it?
- Need to fix the CSS styles that were "fixed" but broke the layout

hjoliver commented:
> Need to fix the CSS styles that were "fixed" but broke the layout

What do you mean by "the layout" here - the graph layout, or more general page layout of the UI?

hjoliver commented:
Technically the "graph layout" is the computed relative positions of all the graph nodes - the shape of the graph, sort of ... so maybe you mean the graph styling (and/or the viewport or whatever the term for that is)

kinow commented:
Oh, yup. Graph styling, sorry. I am just trying to remember where each style is used, and what CSS styles need changing. Should be done by today/tomorrow.

kinow commented on src/components/cylc/graph/Graph.vue:
```diff
@@ -0,0 +1,631 @@
+      class='spinner'
+    ></SyncLoader>
+    <div class='switch-layout ma-0'>
+      <v-expansion-panels
```
Using an expansion panel, I think we have more space to display the graph. It's hidden by default, user can click on the expand button to visualize the settings now.

kinow commented on src/components/cylc/graph/Graph.vue:
Removed data that was not used, or javascript variables declared outside the data (i.e. not reactive, accessible from outside the component), and grouped them with some simple comments for future developers.

kinow commented on src/components/cylc/graph/Graph.vue:
This is the approach of `vue-cytoscape`. We didn't need the whole library, so removed it from `package.json`, and added a few lines here to set up the `<div>` for Cytoscape.

kinow commented on src/components/cylc/graph/Graph.vue:
Spent half day on this one.

The `data` elements in a Vue app are simple data. No promises of methods being accessible or not.

Furthermore, adding a deeply nested object, means Vue needs to iterate it to add proxies/observers.

This call was crashing my browser :open_mouth: By using `Object.freeze`, Vue is later able to tell you don't want reactivity, doesn't bother setting up reactivity, creating proxy objects, etc.

oliver-sanders replied:
Could do with sticking a comment about this in the code (we have a bad history with forgetting important information like this).

kinow replied:
Good idea. Done!

kinow commented on src/components/cylc/graph/Graph.vue:
Also spent a long time searching for this one. Reaaaally bothered by having to always zoom out when opening the graph for my two simple `five` and `families2` workflows :smile:

oliver-sanders replied:
That's a bit nicer than the (100, 100) hack.

kinow commented on src/components/cylc/graph/Graph.vue:
These were not being removed. So we had possibly some detached objects & memory leaking somewhere. (Found about the Panzoom clean up call while looking at their docs to set up default values, by accident).

kinow commented on src/components/cylc/graph/Graph.vue:
GitHub syntax parser bug I think.

kinow commented on src/components/cylc/graph/Graph.vue:
Not quite sure undo-redo is working... but left it anyway. Lumino also has some key listeners...

oliver-sanders replied:
What is is meant to do?

kinow replied:
After you modify the position of nodes, for example, pressing `CTRL + Z` should undo that, and move your node back to its original position.

Demo: https://rawgit.com/iVis-at-Bilkent/cytoscape.js-undo-redo/master/demo.html

kinow commented on src/components/cylc/graph/Graph.vue:
This was tricky too.

Before, we were accessing the element by CSS class.

When using it with Lumino, you would have two widgets & components. The `.cytoscape-navigator-overlay` would return an array with two elements. So only one would be used by both tabs, the other would appear blank.

kinow commented on src/components/cylc/graph/Graph.vue:
Several string concatenation operations replaced by backticks.

However, we still have a lot of `innerHTML` and styles and attributes being changed directly. This is quite bad for performance, possible to cause reflows too.

- https://coderwall.com/p/o9ws2g/why-you-should-always-append-dom-elements-using-documentfragments
- https://www.sitepoint.com/10-ways-minimize-reflows-improve-performance/

oliver-sanders replied:
It's nasty but this code won't make it to 1.0.0 so we can ignore it for now.

kinow commented on src/components/cylc/graph/Graph.vue:
(`which` is deprecated - https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/which, `key` is a possible replacement... TL;DR: `which` behaviour varies per browser)

oliver-sanders replied:
Vue has some support for modifier keys too https://vuejs.org/v2/guide/events.html#System-Modifier-Keys

kinow commented:
I think I got too excited, so didn't squash the commits so that I can revert some if requested. Will add some comments to the new code, but here are the main changes:

- There was some code from `vue-cytoscape` ported to the project.
- Replaced string concatenation in the component by backtick (template literal)
- Moved configuration objects to an `index.js` file
- Moved component and related files to its own folder
- Removed unnecessary files
- Fixed typos and unnecessary or wrong CSS styles
- Added comments to code
- Used a `v-switch` and a `v-select` for layout controls (I think @oliver-sanders suggested it in some issue on github or message on riot :tada: )
- Used a `v-expansion-panel` to display the layout options only when requested (more visible space for the graph)
- Set transparency to the minimap for the same reason above
- Defined initial zoom level, and pan (position of graph)
- Changed styles to use as much viewport as possible (both Graph view and on Lumino as widget)
- Separated DOM elements, so that Lumino widgets are now able to display multiple Graph components
- Added some code to clean up event listeners and Cytoscape objects
- A few more changes here and there where I thought the code could be simpler or the IDE raised some warning

kinow commented on src/components/cylc/graph/Graph.vue:
There were 2 blocks of `css` and `scss`, but `scss` can handle `css` too. The Toppy CSS was imported in the JS part (webpack is able to handle that, but easier to have that here I guess)

oliver-sanders commented on public/index.html:
```diff
@@ -4,7 +4,6 @@
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width,initial-scale=1.0">
     <link href='https://fonts.googleapis.com/css?family=Roboto:300,400,500,700,300italic,400italic' rel="stylesheet" type="text/css">
-    <link href="https://use.fontawesome.com/releases/v5.0.13/css/all.css" rel="stylesheet">
```
FYI we import fontawesome in index.html https://github.com/cylc/cylc-ui/blob/master/public/index.html#L7

kinow replied:
I think you mentioned it somewhere else too. That got me thinking what could be using it, and by coincidence I believe it's only the `Graph` component that was using it for icons.

This code removed in the highlight above is exactly to remove the `public/index.html` import :point_up:

oliver-sanders commented on src/components/cylc/graph/Graph.vue:
Wow I didn't know you could modify an array by messing with its length parameter!

kinow replied:
Couldn't find a more elegant way to clear lists in JS, but I guess that's normal for JS, I think :grin:

kinow commented:
Thanks for the review @oliver-sanders !

kinow commented:
@oliver-sanders sorry, pushed two more commits :point_up: 

The first is to make the widget wrapper to invoke the `resizeGraph()` function of the `Graph` wrapped component, once the widget has been activated. This fixes the Firefox 60.8 issue reported by @hjoliver in #381 

The second commit is to fix a console error I noticed could happen when the user closed the widget while the layout was "running" (`cytoscape.layout().run()` more or less). There is nothing that the user or us can do about it, so the only option for that case is to ignore the error — though it's still possible to capture the error with the browser console by checking the option to break on any exception.

oliver-sanders commented:
> The second commit is to fix a console error I noticed could happen when the user closed the widget while the layout was "running"

Interesting, I didn't see that.

Code looks good.
